### DISCUSSION TITLE
diff2typo

### DISCUSSION THREAD

**Discussion by RainRat** [crate-ci/typos#1038] [Show and tell]
I'm experimenting with a tool that can turn diffs into `typos` data files.
https://github.com/RainRat/diff2typo

1. Find a change that fixed lots of typos.
2. Create a diff for that change, ie. `git diff [parameters] > diff.txt`
3. You can continue with different changes, even different repositories, all into one file ie. `git diff [parameters] >> diff.txt`
4. Run the tool. ie. `python diff2typo.py --input_file diff.txt --output_file
typos.txt --typos_tool_path typos/typos`
5. It will do all the work: extract the corrections, sort, deduplicate, and format it ready for the "Non-critical corrections" thread.

**Comment by RainRat** [comment]
I did a major update:
-splitting compound words
-use dictionary file (to exclude candidates which are more of grammar fixes)
-output format (arrow, csv, table, list)
-improved docs everywhere (code comments, user messages, readme)

**Comment by NathanBaulch** [comment]
FWIW, I usually use the following command to get a rough summary but something more refined sounds good.

```shell
git diff HEAD^! --word-diff-regex='\w+' -U0 \
  | grep -E '\[\-.*\-\]\{\+.*\+\}' \
  | sed -r 's/.*\[\-(.*)\-\]\{\+(.*)\+\}.*/\1 -> \2/' \
  | sort | uniq -c | sort -n
```

**Comment by RainRat** [comment]
**New feature: `diff2typo.py` can output new typos, new corrections for existing typos, or both.**

**New tool: `multitool.py` performs various tasks related to processing one text file into a new one.**

**New tool: `gentypos.py` generates synthetic typos.**

New workflow for using synthetic typos as a basis for finding actual typos.

Possible workflow. Customize steps to your workflow. Mix and match however you like:
 1. Generate list of possible synthetic typos. Use a small wordlist as a basis for the synthetic typos, and a large wordlist for excluding actual valid words. Configure `gentypos.yaml` and run:
   ```bash
   python gentypos.py -c gentypos.yaml
   ```
 2. Use the synthetic typo list with the typos tool to scan your projects. The synthetic typo list will be too large with too many false positives for regular use, but the typos tool can work with it.
   ```bash
typos --config typos_mega.toml --format brief GitHub > typos.txt
   ```
 3. Fix the typos that actually occurred in your project(s)
 4. --- If you don't want to use synthetic typo generation, just start here. ---
 5. Use cmdrunner.py to generate a diff that contains all your typo fixes.
    a. the default cmdrunner.yaml is intended to be run at the root folder of your git projects. Configure cmdrunner.yaml to your needs and run `cmdrunner.py cmdrunner.yaml`
    b. you can concatenate multiple diffs if you want to go back and process multiple past revisions
    c. or, just take a single diff of your project.
 6. Use diff2typo.py to generate a list of typos that actually occurred and you fixed. Make the resulting typo data a permanent part of your workflow, or submit it to the typos maintainer to be added. Example; adjust to your needs.
   ```
diff2typo.py --input_file=diff.txt --typos_tool_path=typos --mode both --dictionary_file=words.csv
   ```
